src/librustc_mir/transform/const_prop.rs

```diff
@@ -560,6 +545,32 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     return None;
                 }
             }
+
+            Rvalue::Aggregate(_, operands) if operands.len() == 0 => {
+                // FIXME(wesleywiser): const eval will turn this into a `const Scalar(<ZST>)` that
+                // `SimplifyLocals` doesn't know it can remove.
+                return None;
+            }
+
+            Rvalue::Aggregate(box AggregateKind::Adt(_, variant_idx, _, _, _), _) => {
+                match &place_layout.variants {
```

Comment:
So my *guess* is that all that code here serves to "guard" against ICEs in the Miri engine? Like, we return early here if we know that `eval_rvalue_into_place` or `eval_place_to_op` would otherwise ICE? I think that should be stated more clearly.